src/differential_arithmetic.jl

```diff
@@ -78,12 +79,34 @@ for T in (:AbstractThunk, :Any)
 end
 
 
-Base.:+(a::AbstractThunk, b::AbstractThunk) = extern(a) + extern(b)
-Base.:*(a::AbstractThunk, b::AbstractThunk) = extern(a) * extern(b)
+Base.:+(a::AbstractThunk, b::AbstractThunk) = unthunk(a) + unthunk(b)
+Base.:*(a::AbstractThunk, b::AbstractThunk) = unthunk(a) * unthunk(b)
 for T in (:Any,)
-    @eval Base.:+(a::AbstractThunk, b::$T) = extern(a) + b
-    @eval Base.:+(a::$T, b::AbstractThunk) = a + extern(b)
+    @eval Base.:+(a::AbstractThunk, b::$T) = unthunk(a) + b
+    @eval Base.:+(a::$T, b::AbstractThunk) = a + unthunk(b)
 
-    @eval Base.:*(a::AbstractThunk, b::$T) = extern(a) * b
-    @eval Base.:*(a::$T, b::AbstractThunk) = a * extern(b)
+    @eval Base.:*(a::AbstractThunk, b::$T) = unthunk(a) * b
+    @eval Base.:*(a::$T, b::AbstractThunk) = a * unthunk(b)
 end
+
+################## Composite ##############################################################
+
+# We intentionally do not define, `Base.*(::Composite, ::Composite)` as that is not meaningful
+# In general one doesn't have to represent multiplications of 2 differentials
+# Only of a differential and a scaling factor (generally `Real`)
+Base.:*(s::Any, comp::Composite) = map(x->s*x, comp)
+Base.:*(comp::Composite, s::Any) = s*comp
```

Comment:
Do we want to assume commutativity here?

Reply:
I feel like we don't need to.
Unlike for `+` where it is super nonsense if not commutative, and making it commutative simplfies code and error message.
For `*` it is not much gain to give up

```suggestion
Base.:*(comp::Composite, s::Any) = map(x->x*s, comp)
```

Longer term I kind of want to restrict to like left multiply by real only or something. 
But that is another PR